django/contrib/auth/hashers.py

```diff
@@ -512,6 +512,82 @@ class BCryptPasswordHasher(BCryptSHA256PasswordHasher):
     digest = None
 
 
+class ScryptPasswordHasher(BasePasswordHasher):
+    """
+    Secure password hashing using the Scrypt algorithm.
+    """
+    algorithm = 'scrypt'
+    block_size = 8
+    maxmem = 0
+    parallelism = 1
+    work_factor = 2 ** 14
+
+    def encode(self, password, salt, n=None, r=None, p=None):
+        assert password is not None
+        assert salt and '$' not in salt
```

Comment:
Is it possible for these asserts (and below) to fail without a bug in Django? If so, proper exceptions should be raised a la: https://code.djangoproject.com/ticket/32508

Reply:
We decided to leave similar asserts in other hashers. Users should not call `encode()` directly.

Reply:
> Users should not call encode() directly.

I looked, and unlike I'm mistaken, Django's documentation does provide an example that instructs users to create a data migration in which the value in user code gets passed to `encode()` as is: https://docs.djangoproject.com/en/3.2/topics/auth/passwords/#password-upgrading-without-requiring-a-login

Reply:
Yes, but these assertions can fail only for users with an incorrect hash in the database. IMO `assert`s can stay here, if you think otherwise we can discuss changing them in all hashers in ticket-32508.

Reply:
For future reference, I added a comment to the ticket here: https://code.djangoproject.com/ticket/32508#comment:35

Reply:
I also commented the same issue.

Reply:
OK, I prepared PR #14680 with a common hook that can be re-used here.

Reply:
Rebased.